toolchain/check/dataflow_analysis.cpp

```diff
@@ -0,0 +1,345 @@
+        auto var_infos = GetVarInfos(sem_ir, assign->lhs_id);
+        for (auto [var_id, _] : var_infos) {
+          facts.assigns.Insert(Fact{inst_id.index, var_id.index});
+        }
+      }
+
+      // 3. Use (NameRef)
+      else if (inst.Is<SemIR::NameRef>()) {
+        auto var_infos = GetVarInfos(sem_ir, inst_id);
+        for (auto [var_id, _] : var_infos) {
+          bool is_lhs = assigned_lhs.Contains(inst_id);
+          // If it's a ref parameter, assignment counts as a use.
+          if (!is_lhs || ref_params.Contains(var_id.index)) {
+            facts.uses.Insert(Fact{inst_id.index, var_id.index});
+          }
+        }
+      }
+
+      // 4. Use (ValueOfInitializer)
+      //    This is used when returning a var by value.
+      else if (auto val_init = inst.TryAs<SemIR::ValueOfInitializer>()) {
+        auto var_infos = GetVarInfos(sem_ir, val_init->init_id);
+        for (auto [var_id, _] : var_infos) {
+          facts.uses.Insert(Fact{inst_id.index, var_id.index});
+        }
+      }
+
+      // 5. Use (AcquireValue)
+      //    This is used when converting a reference to a value (e.g. return
+      //    var).
+      else if (auto acquire = inst.TryAs<SemIR::AcquireValue>()) {
+        auto var_infos = GetVarInfos(sem_ir, acquire->value_id);
+        for (auto [var_id, _] : var_infos) {
+          facts.uses.Insert(Fact{inst_id.index, var_id.index});
+        }
+      }
+
+      // 6. Use (ReturnExpr)
+      //    This is used when returning a var directly (e.g. with return slot).
+      else if (auto ret = inst.TryAs<SemIR::ReturnExpr>()) {
+        auto var_infos = GetVarInfos(sem_ir, ret->expr_id);
+        for (auto [var_id, _] : var_infos) {
+          facts.uses.Insert(Fact{inst_id.index, var_id.index});
+        }
+      }
+
+      // 7. Use (ReturnSlot)
+      //    This is used when a returned var is declared.
+      else if (auto return_slot = inst.TryAs<SemIR::ReturnSlot>()) {
+        auto var_infos = GetVarInfos(sem_ir, return_slot->storage_id);
+        for (auto [var_id, _] : var_infos) {
+          facts.uses.Insert(Fact{inst_id.index, var_id.index});
+        }
+      }
+
+      // 8. Edges (Terminators)
+      if (auto branch = inst.TryAs<SemIR::Branch>()) {
+        facts.branch_edges.Insert(Fact{inst_id.index, branch->target_id.index});
+      } else if (auto branch_if = inst.TryAs<SemIR::BranchIf>()) {
+        facts.branch_edges.Insert(
+            Fact{inst_id.index, branch_if->target_id.index});
+      } else if (auto branch_arg = inst.TryAs<SemIR::BranchWithArg>()) {
+        facts.branch_edges.Insert(
+            Fact{inst_id.index, branch_arg->target_id.index});
+      }
+    }
+  }
+  return facts;
+}
+
+static auto CheckUnusedBindings(Context& context, const DataflowFacts& facts)
+    -> void {
+  auto& sem_ir = context.sem_ir();
+
+  // Collect usage locations. We track the first source-location use for each
+  // variable.
+  llvm::DenseMap<int32_t, SemIR::InstId> first_use;
```

Comment:
Carbon has its own map and set types, and we generally prefer them because they should have higher performance. Had you considered using `Map` here, similar to `Set` above?

Reply:
Done.